### PR TITLE
add bots in diag-app

### PR DESCRIPTION
**Pull request by ledbit**

### CODE REVIEW

File: src/js/app/baseimpl.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ *    this variable will be used to construct the endpoint URLs (NOTE: use singular)
+ * b) a static method _fromApi that processes the API payload and returns an array of objects
+ */
+export default abstract class BaseImpl extends Base {
+    static _type : string = 'unknown';
+    static _fromApi(payload: types.IAPIPayload) : Promise<Array<BaseImpl>> {
+        return Promise.reject('_newObj not implemented');
+    }
+    
+    static load(aid: AssetId) : Promise<Array<BaseImpl>> {
```
Comment (ledbit) [inline src/js/app/baseimpl.ts]: `AssetId` is the correct thing to use here, however it breaks compatibility with the original functions with same name ... :/

File: src/__tests__/backend/space-test.js [hunk 2]
```diff
@@ -127,6 +128,82 @@ describe('App Spaces', () => {
       expect(props.created_at).toBeDefined();
       expect(props._parent).not.toBeDefined();
     });
+
+    // bot stuff 
+
+    it('should not have any bots', () => {
```
Comment (coccyx) [inline src/__tests__/backend/space-test.js]: I have been breaking out tests in `diag-app` as a file per data type, that way they can operate independently. This has some downsides (tests take longer to run etc). Probably needs a complete rethink, but this does break the pattern.

File: src/js/api/datasets.ts [hunk 3]
```diff
@@ -130,3 +130,21 @@ export function getFileContent(sid: string, datasetId: string, fileId: string, o
 export function deleteFile(sid: string, datasetId: string, fileId: string): Promise<types.IAPIPayload> {
   return run('file', baseDelete, joinUri(sid, datasetId, fileId));
 }
+
+///// generic CRUD operations
+
+export function getObjects(type: string, uriParts: Array<string>): Promise<types.IAPIPayload> {
+  return run(type, baseGet, joinUri(...uriParts));
```
Comment (coccyx) [inline src/js/api/datasets.ts]: Very nice. Although your obsession with minimizing boilerplate does make me laugh.

File: src/__tests__/backend/space-test.js (hunk 2 above)
Comment (ledbit) [reply src/__tests__/backend/space-test.js]: yes it does :)

File: src/js/api/datasets.ts (hunk 3 above)
Comment (ledbit) [reply src/js/api/datasets.ts]: what

**Pull request merged**